src/Engine/ProtoCore/Namespace/ElementRewriter.cs

```diff
@@ -0,0 +1,213 @@
+                    // back up on compiler to resolve the namespace from partial name
+                    var matchingClasses = CoreUtils.GetResolvedClassName(classTable, identifier);
+
+                    if (matchingClasses.Length == 1)
+                    {
+                        resolvedName = matchingClasses[0];
+                        var assemblyName = CoreUtils.GetAssemblyFromClassName(classTable, resolvedName);
+
+                        elementResolver.AddToResolutionMap(partialName, resolvedName, assemblyName);
+                    }
+                    else
+                    {
+                        // Namespace conflict, class name could not be resolved.
+                        // This will be reported subsequently in the pre-compilation stage
+                        return;
+                    }
+                }
+                resolvedNames.Enqueue(resolvedName);
+            }
+            
+            RewriteAstWithResolvedName(ref astNode, ref resolvedNames);
+        }
+
+        private static IEnumerable<IdentifierListNode> GetClassIdentifiers(AssociativeNode astNode)
+        {
+            var classIdentifiers = new List<IdentifierListNode>();
+            var resolvedNames = new Queue<string>();
+            DfsTraverse(ref astNode, ref classIdentifiers, ref resolvedNames);
+            return classIdentifiers;
+        }
+
+        /// <summary>
+        /// Replace partial identifier with fully resolved identifier list in original AST
+        /// This is the same AST that is also passed to the VM for execution
+        /// </summary>
+        /// <param name="astNode"></param>
+        /// <param name="resolvedNames"> fully qualified class identifier list </param>
+        private static void RewriteAstWithResolvedName(ref AssociativeNode astNode, ref Queue<string> resolvedNames)
+        {
+            List<IdentifierListNode> classIdentifiers = null;
+            DfsTraverse(ref astNode, ref classIdentifiers, ref resolvedNames);
+        }
+
+        #region private utility methods
+
+        private static void DfsTraverse(ref AssociativeNode astNode, ref List<IdentifierListNode> classIdentifiers, ref Queue<string> resolvedNames)
+        {
+            if (astNode is BinaryExpressionNode)
+            {
+                var bnode = astNode as BinaryExpressionNode;
+                AssociativeNode leftNode = bnode.LeftNode;
+                AssociativeNode rightNode = bnode.RightNode;
+                DfsTraverse(ref leftNode, ref classIdentifiers, ref resolvedNames);
+                DfsTraverse(ref rightNode, ref classIdentifiers, ref resolvedNames);
+
+                bnode.LeftNode = leftNode;
+                bnode.RightNode = rightNode;
+            }
+            else if (astNode is FunctionCallNode)
+            {
+                var fCall = astNode as FunctionCallNode;
+                for (int n = 0; n < fCall.FormalArguments.Count; ++n)
```

Comment:
If this can be turned into `foreach`, please do that. The same treatment should be given to all other `for` loops in this file and make it simpler.